sound/soc/codecs/rt1320-sdw.c

```diff
@@ -2167,7 +2201,7 @@ static int rt1320_sdw_probe(struct sdw_slave *slave,
 	struct regmap *regmap, *mbq_regmap;
 
 	/* Regmap Initialization */
-	mbq_regmap = devm_regmap_init_sdw_mbq(slave, &rt1320_mbq_regmap);
+	mbq_regmap = devm_regmap_init_sdw_mbq_cfg(slave, &rt1320_mbq_regmap, &rt1320_mbq_cfg);
 	if (IS_ERR(mbq_regmap))
 		return PTR_ERR(mbq_regmap);
 
```

Comment:
You should actually be careful here, I am not sure this code is guaranteed to work correctly. This driver registers two register maps using devm, and since it doesn't set the component's regmap using snd_soc_component_init_regmap, then the ASoC core will search for a regmap using dev_get_regmap. Which looks for a regmap devres entry (taking the first it finds), with the two regmaps it is perhaps not entirely clear it will always find the one you want it to.

Obviously if you dropped the second regmap as per my other comment this would be a non-issue, however if you prefer to keep the two regmaps I would suggest calling snd_soc_component_init_regmap from component_probe to avoid any issues.

Reply:
Not following the feedback @charleskeepax. we've used two regmaps for a long time, what I did with this patch was just to shuffle registers between the two regmaps.

If things are broken with two regmaps, then they were broken prior to this patch. D'oh.

Reply:
My point was really that it is quite fragile, because it searches the devres list for the regmap, if you swapped the two devm_regmap_init calls around it would get the other regmap.

Looking through the code most of the controls use custom handlers, I think the only problem you have is "RX Channel Select" before your patch the register used by this control is on the rt1320_sdw_regmap, which is presumably the one the ASoC core finds so it all works. After your patch that register moves to the rt1320_mbq_regmap, which will then presumably not be the one the ASoC core finds. So I would guess, assuming that control has been previously tested, that it worked before your patch but is broken after your patch.

Reply:
I really can't figure out what this component->regmap stuff is supposed to be used for. I only thing I see is this:

````c
void snd_soc_component_setup_regmap(struct snd_soc_component *component)
{
	int val_bytes = regmap_get_val_bytes(component->regmap);

	/* Errors are legitimate for non-integer byte multiples */
	if (val_bytes > 0)
		component->val_bytes = val_bytes;
}
````

but that makes no sense if both the SoundWire and MBQ regmaps can use multiple bytes...


In addition this seems to be used only for a 'bytes' type of control, and I really don't see how this can work with multiple sizes allowed. It makes no sense to rely on the component 'val_bytes' if each control is really different.

IOW, probably the use of two types of regmap per ASoC component probably worked by chance and not by design.

D'oh!

Reply:
component->val_bytes is also used in 

````c
/**
 * snd_soc_get_xr_sx - signed multi register get callback
 * @kcontrol: mreg control
 * @ucontrol: control element information
 *
 * Callback to get the value of a control that can span
 * multiple codec registers which together forms a single
 * signed value in a MSB/LSB manner. The control supports
 * specifying total no of bits used to allow for bitfields
 * across the multiple codec registers.
 *
 * Returns 0 for success.
 */
````

but that really only works if all the registers have the same size...

It feels like the SoundWire and MBQ regmaps exceeed what ASoC assumes.

Reply:
That is what the whole mbq_size callback is handling, basically the regmap pretends to the rest of the world like all the registers are val_bits in size, but may write less bits to the actual hardware if required. This keeps asoc and the regmap cache happy, and ensure we always write the correct amount of data to the hardware.

As for what the component->regmap is used for whenever you define a control or a DAPM widget you generally give it a register, say for example:

static const struct snd_kcontrol_new rt1320_snd_controls[] = {
        SOC_DOUBLE_R_EXT_TLV("FU21 Playback Volume", 
                SDW_SDCA_CTL(FUNC_NUM_AMP, RT1320_SDCA_ENT_FU21, RT1320_SDCA_CTL_FU_VOLUME, CH_01),
                SDW_SDCA_CTL(FUNC_NUM_AMP, RT1320_SDCA_ENT_FU21, RT1320_SDCA_CTL_FU_VOLUME, CH_02),
                0, 0x57, 0, rt1320_set_gain_get, rt1320_set_gain_put, out_vol_tlv),

component->regmap is what the ASoC core uses when it has to access those registers on your behalf, admittedly this is a slightly bad example because that control uses custom callbacks for get and put rather than the core functions but should illustrate the point.

Reply:
Humm, I am not convinced this val_bits/val_bytes works for a 'bytes' array or any computation that uses the maximum value but handles a smaller data set. It's fine for an individual control but for controls that tie multiple registers this seems risky .

Reply:
To be clear are you saying the situation you are concerned about is an MBQ transfer that encompasses multiple registers? I guess I would make two points on that the current MBQ stuff makes no attempt to support transfers over 32-bits, and do we really expect MBQs that encompass multiple registers to be a thing, that seems somewhat like an abuse of the mechanism to me.

Reply:
The problem I worry about is the case where an 'artificial' control is created to e.g. expose L/R from two separate MBQ controls managed by a regmap, using pointer/byte arithmetic to put/get the bytes without knowing that the actual representation is different.

Reply:
I mean if you are twiddling bits I feel like that is on you to make sure it is doing the right thing. But if each of the values are to be transferred as a separate MBQ I see no problem in relation to the current map. You can use regmap_bulk_read/write to access multiple registers at once and the bulk read API is pretty clear it is one register in each val_bits chunk.

Things would perhaps seem a little odd if one was to use regmap_raw_read which should just move bytes on and off the chip. But the nature of the MBQ regmap means it has to use the reg_read/reg_write API rather than the read/write/gather_write API, which means it would be a bug to use the raw operations anyway since they rely on the later API in regmap.

Reply:
I don't have enough understanding of regmap APIs to follow your comments @charleskeepax 

I remain worried by the use of this unique component->val_bytes variable which seems to depend on which regmap gets initialized first and the representation of data in that first regmap.

One day we'll get a nasty side effect by using the wrong representation on the wrong regmap.

Reply:
Its more than component->val_bytes you should worry about there, component->regmap is really the thing that will cause you problems. It will also only get initialised to one of the regmaps and controls all the accesses for ASoC and DAPM. If it gets set to the wrong one never mind it thinking things are the wrong size it just won't see the registers exist.

Reply:
humm, I am not sure if we ever use the component->regmap stuff for now, all the accesses are based on the device->rt7xx->regmap or regmap_mbq pointer.

Reply:
Yes you do, the "RX Channel Select" control will use it, as will the rt1320_spk_l_dac and rt1320_spk_r_dac DAPM widgets and to be fair there is no need for PDE 23 to have a customer handler, so that should also probably be fix and then it would use it too.